workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs

```diff
@@ -23,71 +24,64 @@ internal class SpatialOSBehaviourManager
         private readonly Dictionary<MonoBehaviour, int> numUnsatisfiedReadersOrWriters
             = new Dictionary<MonoBehaviour, int>();
 
-        private readonly Dictionary<MonoBehaviour, Dictionary<uint, IReaderWriterInternal[]>> behaviourToReadersWriters
-            = new Dictionary<MonoBehaviour, Dictionary<uint, IReaderWriterInternal[]>>();
-
-        private readonly Dictionary<uint, HashSet<IReaderWriterInternal>> compIdToReadersWriters =
-            new Dictionary<uint, HashSet<IReaderWriterInternal>>();
-
         private readonly HashSet<MonoBehaviour> behavioursToEnable = new HashSet<MonoBehaviour>();
         private readonly HashSet<MonoBehaviour> behavioursToDisable = new HashSet<MonoBehaviour>();
 
-        private readonly SpatialOSBehaviourLibrary behaviourLibrary;
+        private readonly ReaderWriterStore store;
+        private readonly RequiredFieldInjector injector;
 
         private readonly ILogDispatcher logger;
 
-        private const string LoggerName = "SpatialOSBehaviourManager";
+        private const string LoggerName = nameof(MonoBehaviourActivationManager);
 
-        public SpatialOSBehaviourManager(GameObject gameObject, SpatialOSBehaviourLibrary library, ILogDispatcher logger)
+        public MonoBehaviourActivationManager(GameObject gameObject, RequiredFieldInjector injector,
+            ReaderWriterStore store, ILogDispatcher logger)
         {
             this.logger = logger;
-            behaviourLibrary = library;
+            this.store = store;
+            this.injector = injector;
 
             var spatialComponent = gameObject.GetComponent<SpatialOSComponent>();
             entity = spatialComponent.Entity;
             spatialId = spatialComponent.SpatialEntityId;
 
             foreach (var behaviour in gameObject.GetComponents<MonoBehaviour>())
             {
-                var readerIds = library.GetRequiredReaderComponentIds(behaviour.GetType());
-                foreach (var id in readerIds)
-                {
-                    GetOrCreateValue(behavioursRequiringReaderTypes, id).Add(behaviour);
-                }
+                var readerIds = injector.GetRequiredReaderComponentIds(behaviour.GetType());
+                AddBehaviourForComponentIds(behaviour, readerIds, behavioursRequiringReaderTypes);
```

Comment:
Why do this for behaviours that don't have readers or writers? For the similar reason, why store them in `numUnsatisfiedReadersOrWriters`